File: experiments/haiku-subagents/proxy.py
#!/usr/bin/env python3
"""Logging reverse-proxy for the haiku-subagent harness (curl arms only).

The dotnet servers configure their own logging and suppress ASP.NET request
logs, so HTTP-status friction (302 session-loss, 400 format-guess, 403 rule)
is otherwise invisible — `move_rejected` only captures game-rule rejections
that reach the handler with a valid session. Point agents at this proxy; it
forwards verbatim to the real server and appends one JSONL line per request:
{ts, method, path, status}. Redirects are NOT followed (302 passes through so
the agent behaves exactly as against the real server).

    uv run --no-project experiments/haiku-subagents/proxy.py <listen_port> <target_port> <log_path>
"""
import sys
import json
import time
import http.client
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

LISTEN_PORT = int(sys.argv[1])
TARGET_PORT = int(sys.argv[2])
LOG_PATH = sys.argv[3]

# Hop-by-hop headers must not be forwarded (RFC 7230 §6.1).
HOP_BY_HOP = {
    "connection", "keep-alive", "proxy-authenticate", "proxy-authorization",
    "te", "trailers", "transfer-encoding", "upgrade", "content-length",
}


def _log(method, path, status):
    line = json.dumps({"ts": time.time(), "method": method, "path": path, "status": status})
    with open(LOG_PATH, "a") as f:
        f.write(line + "\n")


class ProxyHandler(BaseHTTPRequestHandler):
    protocol_version = "HTTP/1.1"

    def _proxy(self, method):
        body = b""
        length = int(self.headers.get("Content-Length", 0) or 0)
        if length:
            body = self.rfile.read(length)

        conn = http.client.HTTPConnection("localhost", TARGET_PORT, timeout=30)
        fwd = {k: v for k, v in self.headers.items() if k.lower() not in HOP_BY_HOP}
        try:
            conn.request(method, self.path, body=body, headers=fwd)
            resp = conn.getresponse()
            data = resp.read()
        except Exception as e:
            self.send_error(502, f"proxy error: {e}")
            _log(method, self.path, 502)
            return
        finally:
            conn.close()

        _log(method, self.path, resp.status)
        self.send_response(resp.status)
        for k, v in resp.getheaders():
            if k.lower() in HOP_BY_HOP:
                continue
            self.send_header(k, v)        # replays each Set-Cookie individually
        self.send_header("Content-Length", str(len(data)))
        self.end_headers()
        self.wfile.write(data)

    def do_GET(self):
        self._proxy("GET")

    def do_POST(self):
        self._proxy("POST")

    def log_message(self, *args):
        pass  # silence default stderr access log; we write our own JSONL


if __name__ == "__main__":
    open(LOG_PATH, "w").close()  # fresh log per run
    ThreadingHTTPServer(("localhost", LISTEN_PORT), ProxyHandler).serve_forever()

File: experiments/oss-driver/Proxy.fs
module TicTacToe.OssDriver.Proxy

// Logging reverse-proxy. Forwards verbatim to the real server and appends one JSONL line per
// request ({ts,method,path,status}) — the wire log everything else is graded against. Ported
// from proxy.py. Redirects are NOT followed (302 passes through). Cookies/headers pass through
// unchanged (UseCookies=false), so Set-Cookie and Link/Allow reach the agent exactly as served.
//
//   dotnet run --project experiments/oss-driver -- proxy <listenPort> <targetPort> <logPath>

open System
open System.IO
open System.Net
open System.Net.Http
open System.Text.Json.Nodes
open System.Threading

// Hop-by-hop headers must not be forwarded (RFC 7230 §6.1).
let private hopByHop =
    set [ "connection"; "keep-alive"; "proxy-authenticate"; "proxy-authorization"
          "te"; "trailers"; "transfer-encoding"; "upgrade"; "content-length" ]

let run (argv: string[]) : int =
    if argv.Length < 3 then
        eprintfn "usage: proxy <listenPort> <targetPort> <logPath>"
        2
    else
        let listenPort = int argv.[0]
        let targetPort = int argv.[1]
        let logPath = argv.[2]
        let target = sprintf "http://localhost:%d" targetPort
        File.WriteAllText(logPath, "")  // fresh log per run
        let logLock = obj ()
        let logReq (m: string) (path: string) (status: int) =
            let o = JsonObject()
            o.["ts"] <- JsonValue.Create(float (DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()) / 1000.0)
            o.["method"] <- JsonValue.Create m
            o.["path"] <- JsonValue.Create path
            o.["status"] <- JsonValue.Create status
            lock logLock (fun () -> File.AppendAllText(logPath, o.ToJsonString() + "\n"))

        use client = new HttpClient(new HttpClientHandler(AllowAutoRedirect = false, UseCookies = false))

        let handle (ctx: HttpListenerContext) =
            let req = ctx.Request
            let resp = ctx.Response
            let m = req.HttpMethod
            let rawUrl = req.RawUrl
            try
                use msg = new HttpRequestMessage(HttpMethod(m), target + rawUrl)
                if req.HasEntityBody then
                    use ms = new MemoryStream()
                    req.InputStream.CopyTo ms
                    msg.Content <- new ByteArrayContent(ms.ToArray())
                for key in req.Headers.AllKeys do
                    let kl = key.ToLowerInvariant()
                    if not (hopByHop.Contains kl) && kl <> "host" then
                        let v = req.Headers.[key]
                        if not (msg.Headers.TryAddWithoutValidation(key, v)) && not (isNull msg.Content) then
                            msg.Content.Headers.TryAddWithoutValidation(key, v) |> ignore
                use up = client.Send msg
                let status = int up.StatusCode
                logReq m rawUrl status
                resp.StatusCode <- status
                let relay (h: Headers.HttpHeaders) =
                    for kv in h do
                        let kl = kv.Key.ToLowerInvariant()
                        if hopByHop.Contains kl then ()
                        elif kl = "content-type" then resp.ContentType <- String.Join(",", kv.Value)
                        else for v in kv.Value do (try resp.Headers.Add(kv.Key, v) with _ -> ())  // Set-Cookie replayed individually
                relay up.Headers
                relay up.Content.Headers
                let data = up.Content.ReadAsByteArrayAsync().Result
                resp.ContentLength64 <- int64 data.Length
                resp.OutputStream.Write(data, 0, data.Length)
            with e ->
                (try logReq m rawUrl 502 with _ -> ())
                (try resp.StatusCode <- 502 with _ -> ())
            (try resp.OutputStream.Close() with _ -> ())

        let listener = new HttpListener()
        listener.Prefixes.Add(sprintf "http://localhost:%d/" listenPort)
        listener.Start()
        eprintfn "proxy: localhost:%d -> localhost:%d (log %s)" listenPort targetPort logPath
        while true do
            let ctx = listener.GetContext()
            ThreadPool.QueueUserWorkItem(fun _ -> handle ctx) |> ignore
        0
